Spec.md

```diff
@@ -0,0 +1,223 @@
+
+## Requirements for the Mapping of the address space to PubSub
+
+(Reference: [umati/UA-CloudPublisher#5](https://github.com/umati/UA-CloudPublisher/issues/5))
+
+- Ability to react to dynamic changes in the address space (nodes added/deleted)
+  - Changes should be detected by monitoring NodeManagement events.
+- Resolve references to nodes (NodeId as values)
+- Hierarchy references must be mapped
+- Non-hierarchy references should be mapped if necessary
+  - Non-hierarchy references are mapped if relevant to the data semantics.
+- Provide type information
+- Methods should be considered
+- Online detection (possibly as separate topics)
+- Allow empty objects
+- Events/alarms must be mappable
+- Coordinate the completeness of the run
+- Encoding should be UA JSON
+- Dimensioning for approximately 350 subscribers (machines + dashboard users)
+- Custom DataTypes must be included
+- The address space includes at least OPC UA Machinery and possibly other Companion Specifications based on it.
+- The entry point of the mapping is the machine object in the Machines folder of Machinery.
+
+## Design Decisions
+
+The concept of OPC UA Part 14 should be followed as closely as possible.
+
+## Mapping
+
+### Mapping of Objects and Variables
+
+- An object is mapped to a DataSet.
+- A variable is mapped as a field to the DataSet. If the variable contains children (other variables), a DataSet with the children is created.
+
+#### Example
+
+![image](https://github.com/umati/UA-CloudPublisher/assets/12342602/91eae1d7-9d6f-4246-abb2-0ec5ea9b59bc)
+
+The image shows a part of an address space. Each object is mapped to a DataSet:
+
+- Production
+- ActiveProgram
+- State
+
+The properties Name, NumberInList, Id, and Number are fields of the DataSets. The value of CurrentState is mapped as a field of ActiveProgram and has a DataSet with its properties:
+
+- Production
+- ActiveProgram
+  - Name
+  - NumberInList
+  - CurrentState
+- State
+- CurrentState
+  - Id
+  - Number
+
+### Explanation
+
+A field can also contain properties but no other variables, even though the address space can contain more hierarchies. Properties of objects also cannot be mapped to the properties of a field. Therefore, a uniform mapping is used for objects and variables so that access is always consistent.
+
+## Mapping of References
+
+The name of a DataSet is created by the BrowsePath of the object. The NamespaceUri is used instead of the NamespaceIndex to make the path server-independent. Other BrowsePaths may also be possible. Alternative BrowsePaths can be mapped to the description (see description field of the MetaData). References can also be mapped into the description.
+
+## Topic Structure
+
+The generic topic structure for OPC UA is:
+
+`<Prefix>/<Encoding>/<MqttMessageType>/<PublisherId>/[<WriterGroup>[/<DataSetWriter>]]`
+
+For the umati Dashboard, this topic structure is restricted as follows:
+
+- `<Prefix> = umati.v3`

```

Comment:
umativ3

Reply:
I think a seperate char is good. maybe umati_v3?

Reply:
umati/v3

Reply:
this results in the problem that OPC UA Part 14 suggest only one level as prefix.

Reply:
>The <Prefix> should be one [Topic](https://reference.opcfoundation.org/search/362?t=Topic) level, however, system designers may break <Prefix> into multiple [Topic](https://reference.opcfoundation.org/search/362?t=Topic) levels. Note that using multiple [Topic](https://reference.opcfoundation.org/search/362?t=Topic) levels prevents [Subscribers](https://reference.opcfoundation.org/search/362?t=Subscribers) from automatically discovering [Publishers](https://reference.opcfoundation.org/search/362?t=Publishers) in the system unless they are preconfigured with the <Prefix> used for the system.
> https://reference.opcfoundation.org/Core/Part14/v105/docs/7.3.5.7

It may only break some autodiscovery, which propably does not exist at the moment.

Reply:
@GoetzGoerisch what do you think? Should we split the prefix?

Reply:
Yes, we should split and use two levels.
Please keep in mind, that the custom topic tree needs to be published in the connection topic.
[ `https://reference.opcfoundation.org/Core/Part14/v105/docs/6.4.2.5.7`]( https://reference.opcfoundation.org/Core/Part14/v105/docs/6.4.2.5.7)

Reply:
done